src/ISDeviceCal.cpp

```diff
@@ -0,0 +1,822 @@
+struct sOrthoCal;
+struct sCalData;
+
+// Macro for checking if vector is all zeros
+#define VEC3_ALL_ZERO(v) ((v)[0] == 0.0f && (v)[1] == 0.0f && (v)[2] == 0.0f)
+
+// Helper function prototypes
+static std::vector<std::string> split(const std::string& s, char delimiter);
+static std::string getFilename(const std::string& path);
+static std::string getParentPath(const std::string& path);
+static bool createDirectoryRecursive(const std::string& path);
+static bool mat3x3_IsIdentity(const float* mat);
+// static uint32_t flashChecksum32(const void* data, int size);
+static std::string formatString(const char* format, ...);
+static void logDebug(const char* format, ...);
+
+// Helper function implementations
+static std::vector<std::string> split(const std::string& s, char delimiter) {
+    std::vector<std::string> tokens;
+    std::string token;
+    std::istringstream tokenStream(s);
+    while (std::getline(tokenStream, token, delimiter)) {
+        tokens.push_back(token);
+    }
+    return tokens;
+}
+
+static std::string getFilename(const std::string& path) {
+    size_t pos = path.find_last_of("/\\");
+    if (pos != std::string::npos) {
+        return path.substr(pos + 1);
+    }
+    return path;
+}
+
+static std::string getParentPath(const std::string& path) {
+    size_t pos = path.find_last_of("/\\");
+    if (pos != std::string::npos) {
+        return path.substr(0, pos);
+    }
+    return ".";
+}
+
+static bool createDirectoryRecursive(const std::string& path) {
+    if (path.empty()) return true;
+    
+    struct stat info;
+    if (stat(path.c_str(), &info) == 0 && (info.st_mode & S_IFDIR)) {
+        return true; // Directory already exists
+    }
+    
+    // Create parent directory first
+    std::string parent = getParentPath(path);
+    if (!parent.empty() && parent != "." && parent != path) {
+        if (!createDirectoryRecursive(parent)) {
+            return false;
+        }
+    }
+    
+    // Create this directory
+    return mkdir(path.c_str(), 0755) == 0 || errno == EEXIST;
+}
+
+static bool mat3x3_IsIdentity(const float* mat) {
+    const float epsilon = 1e-6f;
+    for (int i = 0; i < 3; i++) {
+        for (int j = 0; j < 3; j++) {
+            float expected = (i == j) ? 1.0f : 0.0f;
+            if (fabs(mat[i * 3 + j] - expected) > epsilon) {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+// static uint32_t flashChecksum32(const void* data, int size) {
+//     // Simple checksum implementation
+//     const uint8_t* bytes = (const uint8_t*)data;
+//     uint32_t checksum = 0;
+//     for (int i = 0; i < size; i++) {
+//         checksum += bytes[i];
+//     }
+//     return checksum;
+// }
+
+static void logDebug(const char* format, ...) {
+    va_list args;
+    va_start(args, format);
+    vprintf(format, args);
+    va_end(args);
+}
+
+static std::string formatString(const char* format, ...) {
```

Comment:
utils.util.cpp has a utils::string_format() function which already provides this capability. It should be used instead.